### ISSUE TITLE
RPC Services

### ISSUE THREAD

**Issue by alexbiehl** [k-bx/protocol-buffers#9]
Hey, 

I got handed a bunch of .proto files using rpc service definitions. They change quite frequently and I don't like to manually keep the bindings up-to-date so I am planing to invest some time into implementing the generation. 

The obvious question is how to encode the services. One could do it with typeclasses, like the thrift package does, or encode in some data type, maybe also GADTs which would be nice. 

@k-bx: Have you had any thought on that? Is there anything which would prevent the work to implement it?

**Comment by k-bx** [comment]
@alexbiehl not sure what exactly do you have on your mind, so please feel free to move forward, we'll review and discuss the work then.

**Comment by alexbiehl** [comment]
I have put together a patch: https://github.com/alexbiehl/protocol-buffers
It turns 
````

message SearchRequest {
    required string query = 1;
    optional int32 page_number = 2;
    optional int32 result_per_page = 3;
}

message SearchResponse {
    repeated string results = 1;
}

service SearchService {
    rpc Search (SearchRequest) returns (SearchResponse);
}
````

into something like this

````
type SearchService = '[Search]
 
type Search = P'.Method ".Service.SearchService.Search" SearchRequest SearchResponse
 
search :: Search
search = P'.Method
````

As protocol-buffers is agnostic to the specific rpc backend. It makes sense to generate a generic specification for the service. This is done here in a servant like way. An implementor just specifies how to interpret the types. The implementor is not being constraint by some shape of type class. The implementor chooses. Additionally we can easily derive servers for RPC services. 

Would you be interested in that?

**Comment by wavewave** [comment]
Great! By the way, recently, I have implemented oneof (See issue #4 and pull request #19). It looks like merging your patch with pull request #19 is not difficult fortunately (not guaranteed though). Thanks for your contribution.

**Comment by wavewave** [comment]
@alexbiehl Looking into what the above comment suggests more, it's very cool to use type-level list for defining service interface. It will help to add extra type-safety to haskell side service definition.

One thing that I think useful is to generate a data type that contains all the methods for a given service (i.e. allowing for value-level inspection) . Such container data type may be necessary for dispatching events dynamically through a given service interface.  That's probably what you mean by GADTs in the first comment.

**Comment by wavewave** [comment]
Maybe with some helper functions, we can obtain representing value for each service.
https://hackage.haskell.org/package/typelits-witnesses-0.1.1.0/docs/GHC-TypeLits-List.html

**Comment by alexbiehl** [comment]
Glad you liked it. 

A GADT approach also needs a type class to associate method names to the constructors. We could also try to wrap the methods in some existencial quantified data constructor: 

````
data SomeMethod m = forall name req resp . (KnownSymbol name, Wire req, Wire resp) => name (req -> m resp)
````
However I am not sure how useful it would be.

**Comment by wavewave** [comment]
@alexbiehl Thanks. That generic GADT may be all we need, and then we do not have to think about it any more on the code generation side. Anyway, for making such handler, a user need to write each handling function by hand, so it's more likely to be the user's job.

**Comment by wavewave** [comment]
@alexbiehl I've merged your `codegen-services` and my `oneof` successfully!
See the resultant branch. 
https://github.com/wavewave/protocol-buffers/tree/oneof-service

I have tested service code generation with the above search service example. By the way, I made missing pragma DataKinds added into generated module, and also correctly qualify type name in service module. otherwise, generated code didn't compile.

Once #19 is merged, we will be able to make another pull request for this.

**Comment by k-bx** [comment]
All right, #19 is merged.

**Comment by wavewave** [comment]
@k-bx Thanks!  I am going to add some plausible example and some explanation about this service support into README.md, and make a pull request. The code itself seems to work well at least. :-)

**Comment by alexbiehl** [comment]
I have some sketches for an imaginery client and server to show how its supposed to be used which ill post later.

**Comment by alexbiehl** [comment]
and i also implemented a hadoop rpc client as proof of concept which we can refer to!

**Comment by alexbiehl** [comment]
I have put together a little example of how I intend to use the definitions here:
https://gist.github.com/alexbiehl/60b07a7f7a0074970fe6

**Comment by wavewave** [comment]
Super!

**Comment by tvh** [comment]
Implemented in https://github.com/k-bx/protocol-buffers/pull/96
